enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,442 @@
+---
+title: aws-pod-identity 
+authors:
+  - "@sjenning"
+reviewers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+approvers:
+  - "@derekwaynecarr"
+creation-date: 2020-03-26
+last-updated: 2020-03-26
+status: provisional
+see-also:
+  - ""  
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# Support Native (IAM-based) Pod Identity on AWS
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+When deployed on AWS, Kubernetes can use pod identity managed by IAM.  AWS has a
+[webhook](https://github.com/aws/amazon-eks-pod-identity-webhook/) they maintain
+for EKS that does enables this ability.  This enhancement integrates the same
+functionality into OpenShift.
+
+The `cloud-credential-operator` will be responsible for managing this new
+ability as it serves as an alternative credential _injection_ mechanism versus
+the existing mechanism of injecting an IAM user acccess ID and key into a
+secret.  The existing role/user _creation_ via a CredentialRequest CR is common
+between the methods.
+
+The new injection mechanism works by creating a `ServiceAccount` annotated with
+a IAM role the pod should assume. When a pod runs with the annoated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contanis the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
+   bounded token
+2. An admin can create a IAM role via `CredentialRequest` CR and inject to a
+   `ServiceAccount` (just as they can create an IAM user and inject to a
+   `Secret` currently)
+3. OpenShift should create the OIDC provider in IAM
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seemsless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
```

Comment:
this will be a follow-on enhancement.

Reply:
So the enhancement proposes using Mutation webhook for pods to add the envs and projected volumes.

1. That means that core components cannot create pods until this webhook server is running.
2. The pods cannot be modified day-2 to add these envs, projected volumes due to restriction on pods. (only image is allowed to be changed in an existing pod)
3. That means there is going to be a requirement for a `force roller` of deployments/daemonsets/jobs etc that will roll these pods to have new pods created that can be affected by the webhook
4. what about static pods, kube-controller-manager is a static pod today how will be provide/affect that pod.
5. what about networking pods that would also end up requiring the mutation, but bootstrap-control-plane doesn't have capability to run any webhooks

even though it's not the goal of this enhancement to support this for core components, picking and using webhooks has effect on the long term goal. so I think we should think about it here.

Reply:
we don't have to implement all, but proposing the change in phase, phase 1: user workload, phase 2: core components